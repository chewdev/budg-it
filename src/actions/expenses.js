import {
  database,
  ref,
  get,
  child,
  push,
  set,
  remove,
  update,
} from "../firebase/firebase";

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    const userExpensesRef = ref(database, `users/${uid}/expenses`);
    const newExpenseRef = push(userExpensesRef);

    return set(newExpenseRef, expense).then((ref) => {
      dispatch(
        addExpense({
          id: newExpenseRef.key,
          ...expense,
        })
      );
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenseToRemoveRef = ref(database, `users/${uid}/expenses/${id}`);
    return remove(expenseToRemoveRef).then(() => {
      dispatch(
        removeExpense({
          id,
        })
      );
    });
  };
};

export const removeExpensesChosen = (ids = []) => ({
  type: "REMOVE_EXPENSES_CHOSEN",
  ids,
});

export const startRemoveExpensesChosen = (ids = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    ids.forEach((id) => {
      const expenseToRemoveRef = ref(database, `users/${uid}/expenses/${id}`);
      return remove(expenseToRemoveRef).then(() => {
        dispatch(removeExpense({ id }));
      });
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenseToUpdateRef = ref(database, `users/${uid}/expenses/${id}`);
    return update(expenseToUpdateRef, updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return get(child(ref(database), `users/${uid}/expenses`)).then(
      (snapshot) => {
        const expenses = [];

        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      }
    );
  };
};
