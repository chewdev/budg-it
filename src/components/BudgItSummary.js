import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectIncome from '../selectors/income';
import selectTotal from '../selectors/total';

export const BudgItSummary = (props) => {
    const expenseWord  = props.expenses.length === 1 ? 'expense ' : 'expenses ';
    const incomeWord = props.income.length === 1 ? 'income ' : 'incomes ';
    const budgitTotal = selectTotal(props.expenses, props.income);
    return (
    <div className={budgitTotal >= 0 ? "page-header" : "page-header page-header--bg-red"}>
        <div className="content-container">
            <h2 className="page-header__budgit">
                Budget: <span className={budgitTotal === 0 ? "page-header__brown" : budgitTotal > 0 ? "page-header__main-green" : "page-header__red"}>{numeral(budgitTotal).format('$0,0.00')}</span>
            </h2>
            <div className="page-header__sub-summary">
                <h3 className="page-header__expenses">
                    Viewing <span className="page-header__expense-number">{props.expenses.length}</span> {expenseWord}
                    totaling: <span className="page-header__expense-total">{numeral(selectTotal(props.expenses)).format('$0,0.00')}</span>
                </h3>
                <h3 className="page-header__incomes">
                    Viewing <span className="page-header__income-number">{props.income.length}</span> {incomeWord}
                    totaling: <span className="page-header__income-total">{numeral(selectTotal(props.income)).format('$0,0.00')}</span>
                </h3>
            </div>
            <div className="page-header__sub-summary">
                <Link to='/create/expense' className="btn btn--expense">Add Expense</Link>
                <Link to='/create/income' className="btn btn--income">Add Income</Link>
            </div>
        </div>
    </div>
)};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
        income: selectIncome(state.income, state.filters)
    };
};

export default connect(mapStateToProps)(BudgItSummary);