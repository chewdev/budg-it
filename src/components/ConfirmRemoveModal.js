import React from "react";
import Modal from "react-modal";

const ConfirmRemoveModal = (props) => {
  Modal.setAppElement("body");
  return (
    <Modal
      contentLabel="Confirm Remove Modal"
      isOpen={props.isOpen}
      onRequestClose={props.handleCloseModal}
      closeTimeoutMS={200}
      className="modal"
    >
      <h3 className="modal__title">
        Are you sure you want to remove all {props.itemType}?
      </h3>
      <button
        className="btn btn--confirm-remove"
        onClick={props.onConfirmRemoveAllVis}
      >
        Confirm Removal
      </button>
      <button
        className="btn btn--cancel-remove"
        onClick={props.handleCloseModal}
      >
        Cancel
      </button>
    </Modal>
  );
};

export default ConfirmRemoveModal;
