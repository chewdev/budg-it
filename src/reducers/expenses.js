// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'REMOVE_EXPENSES_CHOSEN':
            action.ids.forEach((id) => {
            state = state.filter((expense) => expense.id !== id);
            });
            return state;
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        case 'SET_EXPENSES':
            return action.expenses;
        case 'REMOVE_ALL':
            return [];
        default:
            return state;
    }
};