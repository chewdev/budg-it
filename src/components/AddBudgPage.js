import React from 'react';
import BudgForm from './BudgForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddBudgPage extends React.Component {
	onSubmit= (expense) => {
		// props.dispatch(addExpense(expense));
		this.props.startAddExpense(expense);
		this.props.history.push('/');
	};
	render() {
		return (
			<div>
				<div className="page-header page-header--bg-red">
					<div className="content-container">
						<h2 className="page-header__title page-header__red">Add Expense</h2>
					</div>
				</div>
				<div className="content-container">
					<BudgForm 
						buttonText="Add Expense"
						onSubmit={this.onSubmit}
					/>
				</div>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({startAddExpense: (expense) => dispatch(startAddExpense(expense))});
	

export default connect(undefined, mapDispatchToProps)(AddBudgPage);