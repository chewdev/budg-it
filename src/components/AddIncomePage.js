import React from 'react';
import BudgForm from './BudgForm';
import { connect } from 'react-redux';
import { startAddIncome } from '../actions/income';

export class AddIncomePage extends React.Component {
	onSubmit = (income) => {
		this.props.startAddIncome(income);
		this.props.history.push('/');
	};
	 render() { 
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h2 className="page-header__title page-header__main-green">Add Income</h2>
					</div>
				</div>
				<div className="content-container">
					<BudgForm 
						buttonText="Add Income"
						onSubmit={this.onSubmit}
					/>
				</div>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({startAddIncome: (income) => dispatch(startAddIncome(income))});

export default connect(undefined, mapDispatchToProps)(AddIncomePage);