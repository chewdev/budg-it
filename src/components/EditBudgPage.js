import React from 'react';
import { connect } from 'react-redux';
import BudgForm from './BudgForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditBudgPage extends React.Component {
	onSubmit = (expense) => {
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	};
	onRemove = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push('/');
	}
	render() {
		return (
			<div>
				<div className="page-header page-header--bg-red">
					<div className="content-container">
						<h2 className="page-header__title red">Edit Expense</h2>
					</div>
				</div>
				<div className="content-container">
					<BudgForm
						buttonText={"Save Expense"}
						expense={this.props.expense}
						onSubmit={this.onSubmit}
					/>
					<button
						className="btn btn--expense btn--remove" 
						onClick={this.onRemove}
					>
						Remove Expense
					</button>
				</div>
			</div>
			);
	}
};

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find((expense) => expense.id === props.match.params.id)
	};
};

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditBudgPage);