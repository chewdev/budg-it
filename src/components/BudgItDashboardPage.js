import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import BudgItSummary from './BudgItSummary';
import IncomeList from './IncomeList';

const BudgItDashboardPage = () => (
	<div>
		<BudgItSummary />
		<ExpenseListFilters />
		<ExpenseList />
		<IncomeList />

	</div>
);

export default BudgItDashboardPage;