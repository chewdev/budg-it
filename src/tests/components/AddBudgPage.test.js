import React from 'react';
import { shallow } from 'enzyme';
import { AddBudgPage } from '../../components/AddBudgPage';
import expenses from '../fixtures/expenses';

let addExpense, history, wrapper;

beforeEach(() => {
    addExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddBudgPage addExpense={ addExpense } history={ history } />);
});

test('should render AddBudgPage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
    wrapper.find('BudgForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpense).toHaveBeenLastCalledWith(expenses[1]);
});