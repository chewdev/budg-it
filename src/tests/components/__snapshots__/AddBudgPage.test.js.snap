// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddBudgPage correctly 1`] = `
<div>
  <div
    className="page-header page-header--bg-red"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title page-header__red"
      >
        Add Expense
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <BudgForm
      buttonText="Add Expense"
      onSubmit={[Function]}
    />
  </div>
</div>
`;
